consensus/raft/raft.go

```diff
@@ -427,9 +435,134 @@ func (rw *raftWrapper) Peers() ([]string, error) {
 	return ids, nil
 }
 
-// only call when Raft is shutdown
-func (rw *raftWrapper) Clean() error {
-	dbh := newDataBackupHelper(rw.dataFolder)
+// latestSnapshot looks for the most recent raft snapshot stored at the
+// provided basedir.  It returns a boolean indicating if any snapshot is
+// readable, the snapshot's metadata, and a reader to the snapshot's bytes
+func latestSnapshot(raftDataFolder string) (bool, *hraft.SnapshotMeta, io.ReadCloser, error) {
+	store, err := hraft.NewFileSnapshotStore(raftDataFolder, RaftMaxSnapshots, nil)
+	if err != nil {
+		return false, nil, nil, err
+	}
+	snapMetas, err := store.List()
+	if err != nil {
+		return false, nil, nil, err
+	}
+	if len(snapMetas) == 0 { // no error if snapshot isn't found
+		return false, nil, nil, nil
+	}
+	meta, r, err := store.Open(snapMetas[0].ID)
+	if err != nil {
+		return false, nil, nil, err
+	}
+	return true, meta, r, nil
+}
+
+// LastState decodes the latest snapshot into the provided state
+func LastState(cfg *Config, state state.State) (bool, error) {
+	// Read most recent snapshot
+	dataFolder, err := makeDataFolder(cfg.BaseDir, cfg.DataFolder)
+	if err != nil {
+		return false, err
+	}
+	found, _, r, err := latestSnapshot(dataFolder)
+	if err != nil {
+		return false, err
+	}
+	if !found {
+		return false, nil
+	}
+
+	// !!!!!!Use Decodesnapshot instead of unmarshal
+	rawBytes, err := ioutil.ReadAll(r)
+	if err != nil {
+		return true, err
+	}
+	err = p2praft.DecodeSnapshot(rawBytes, state)
+	if err != nil {
+		return true, err
+	}
+	return true, nil
+}
+
+// SnapshotMigrate saves a raft snapshot containing a migrated version of the
+// latest snapshot's state.  If no state yet exists there is nothing to migrate
+// and it returns an error.  Old snapshots are cleaned up.
+// Only call when Raft is shutdown.
+func SnapshotMigrate(cfg *Config, newState state.State) error {
+	dataFolder, err := makeDataFolder(cfg.BaseDir, cfg.DataFolder)
+	if err != nil {
+		return err
+	}
+	// Get info from latest snapshot
+	found, meta, r, err := latestSnapshot(dataFolder)
+	if err != nil {
+		return err
+	}
+	if !found {
+		return errors.New("No existing snapshot to migrate from")
+	}
+
+	// migrate state
+	oldStateBytes, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	err = p2praft.DecodeSnapshot(oldStateBytes, newState)
+	if err != nil {
+		logger.Debug("error decoding snapsot")
+		return err
+	}
```

Comment:
the code here and up is very similar to LastState()

Reply:
Will be fixed while addressing previous comment